cluster-up/cluster/k8s-1.17.0/dev-guide.md

```diff
@@ -0,0 +1,86 @@
+flow of start a pre-provisioned cluster, flow of creating a new provider, and how to create a new provider.
+
+A provider includes all the images (K8s base image, nodes OS image) and the scripts that allow to start a
+cluster offline, without downloading nor installing or compiling new resources.
+
+# Project structure
+* cluster-provision folder - creating preprovisioned clusters.
+* cluster-up folder - spinning preprovisioned clusters.
+* gocli - gocli is a binary that assist in provisioning and spinning a cluster.
+
+# k8s Deployment
+Running `make cluster-up` will deploy a pre-provisioned cluster.
+Upon finishing deployment of a K8s deploy, we will have 3 containers:
+* k8s-1.17.0 vm container - a container that runs a qemu VM, which is the K8s node.
+* Registry container - a shared image registry.
+* k8s-1.17.0 dnsmasq container - a container that run dnsmasq, which gives dns and dhcp services.
+
+The containers are running and looks like this:
+```
+[root@modi01 1.17.0]# docker ps
+CONTAINER ID        IMAGE                   COMMAND                  CREATED             STATUS              PORTS                                                                                                                          NAMES
+3589e85efc7d        kubevirtci/k8s-1.17.0   "/bin/bash -c '/vm.s…"   About an hour ago   Up About an hour                                                                                                                                   k8s-1.17.0-node01
+4742dc02add2        registry:2.7.1          "/entrypoint.sh /etc…"   About an hour ago   Up About an hour                                                                                                                                   k8s-1.17.0-registry
+13787e7d4ac9        kubevirtci/k8s-1.17.0   "/bin/bash -c /dnsma…"   About an hour ago   Up About an hour    127.0.0.1:8443->8443/tcp, 0.0.0.0:32794->2201/tcp, 0.0.0.0:32793->5000/tcp, 0.0.0.0:32792->5901/tcp, 0.0.0.0:32791->6443/tcp   k8s-1.17.0-dnsmasq
+```
+
+Nodes:
+```
+[root@modi01 kubevirtci]# oc get nodes
+NAME     STATUS   ROLES    AGE   VERSION
+node01   Ready    master   83m   v1.17.0
+```
+
+# Flow of K8s provisioning (1.17 for example)
+`cluster-provision/k8s/1.17.0/provision.sh`
+* Runs the common cluster-provision/k8s/provision.sh.
+    * Runs cluster-provision/cli/cli (bash script).
+        * Create a container, and runs vm.sh in it which creates a vm using qemu, and checks its ready (according ssh).
+        * Runs ./cluster-provision/k8s/scripts/provision.sh in the container.
+        * Shutdown the vm.
+
+# Flow of K8s cluster-up (1.17 for example)
+Run
+```
+export KUBEVIRT_PROVIDER=k8s-1.17.0
+make cluster-up
+```
+* Runs ./cluster-up/up.sh which sources the following.
+    * ./cluster-up/cluster/k8s-1.17.0/provider.sh (selected according $KUBEVIRT_PROVIDER)
+    * ./cluster-up/cluster/k8s-provider-common.sh
```

Comment:
didn't find it in the script. do you mean
source ${KUBEVIRTCI_CLUSTER_PATH}/$KUBEVIRT_PROVIDER/provider.sh
?

Reply:
alias f='find . -name'

[root@modi01 kubevirtci]# f k8s-provider-common.sh
./cluster-up/cluster/k8s-provider-common.sh

cluster-up/cluster/k8s-1.17.0/provider.sh:source ${KUBEVIRTCI_PATH}/cluster/k8s-provider-common.sh
which resolves to hardcoded 
     * ./cluster-up/cluster/k8s-provider-common.sh

Reply:
ok, but if I am not mistaken, cluster-up/up.sh:11 is 
source ${KUBEVIRTCI_PATH}hack/common.sh
how is it resolved to: ./cluster-up/cluster/k8s-provider-common.sh ?

Reply:
i meant that 
   * ./cluster-up/cluster/k8s-1.17.0/provider.sh (selected according $KUBEVIRT_PROVIDER)
sources the next line 
   * ./cluster-up/cluster/k8s-provider-common.sh

uploading a chance soon and lets see there whats the status